task/gpx_ux/centralize_map_on_cesium_ip

BarAgam opened the pull request:
Started from adding a simple "centralize map on israel on initializtion"
I added "on-ready" event for GGMapView that the ViewerFragment is registeration.
Refactored code so that no more JS code runs untill CesiumMapView is ready.

barmako committed:
JS notifies JAVA on cesium ready

barmako committed:
Implemented JS zoomToRectangle function.
Added a listener interface to when MapView is ready for action.
Added a setter for listener interface at GGMapView interface.
Added a getter for GGMapView initialized state.
Implemented that for CesiumMapView.

ViewerFragment reorganized (methods re-ordering by access modifiers)
ViewerFragment implements said interface and:
    Zooms to Israel~ extent when cesium has done loading.
    Attaches SentLocationLayer and UserLocationLayer to GGMapView
    Registers for location updates through receivers

review-ninja commented:
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/24" target="_blank"><img src="https://app.review.ninja/52010471/pull/24/badge" alt="ReviewNinja"/>

teamagam1 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -58,7 +58,7 @@
         </activity>
 
         <service
-            android:name=".app.network.services.GGMessageLongPollingService"
```
מה קרה פה? 
זה כאן כדי להשאר?

BarAgam replied:
שורות שקולות
לדעתי, לא משנה מה נבחר להשאיר
תחליט

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/GGApplication.java:
```diff
@@ -15,48 +15,12 @@
-    }
 
     /**
      * Checks if the current version is increased since the last version that was saved in mPrefs.
```
ההערה שייכת לקוד שנמחק. למחוק גם את ההערה. !fix

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/network/services/GGMessageLongPollingService.java:
```diff
@@ -92,7 +92,11 @@ protected void onHandleIntent(Intent intent) {
             final String action = intent.getAction();
             if (ACTION_MESSAGE_POLLING.equals(action)) {
-                mPoller.poll();
+                try {
```
מה הקשר עכשיו? 
למה הוספת את זה? אולי תרצה להגדיר Exception 
ספציפי?

BarAgam replied:
זה קוד שאם לא נעטוף יכול להעיף את האפליקציה.
מה גם שזה קוד "מועד" לבעיות בגלל המשימה שלו.
אם בעתיד נגדיר שגיאה מיוחד, בעתיד נטפל בה בנפרד

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -64,6 +65,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         mUsersLocationsLayer = new VectorLayer("vlUsersLocation");
 
         mGGMapView = (GGMapView) rootView.findViewById(R.id.gg_map_view);
+        mGGMapView.setOnReadyListener(this);
```
האם יש סיכוי שהצזיום יסיים לעלות לפני ששורת הקוד הזאת תקרא? ואז לא נתפוס אף פעם את הקריאה מהצזיום

BarAgam replied:
כן, זאת בעיה שתאורטית משמידה את האפליקציה

BarAgam replied:
וטופלה. !resolved

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -176,15 +138,74 @@ public void onDetach() {
+    @Override
+    public void onGGMapViewReady() {
+        //Zoom to Israel~
+        mGGMapView.setExtent(34, 29, 36, 34);
```
you could use here some values file (res/values/arrays.xm?).

.

BarAgam replied:
!resolved

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -176,15 +138,74 @@ public void onDetach() {
+        registerForLocationUpdates();
+    }
+
+    private void registerForLocationUpdates() {
```
what happens with life cycle?
onRestart? onResume? does this function will be called again?

BarAgam replied:
As of now, the whole JS part of our viewer life-cycle behavior is unclear.
This is a real problem we'll need to tackle in the near future, but it doesn't concern this part of the code alone

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -176,15 +138,74 @@ public void onDetach() {
         super.onDestroyView();
-        MessageBroadcastReceiver.unregisterReceiver(getActivity(), mUserLocationReceiver);
-        mLocationFetcher.unregisterReceiver(mLocationReceiver);
+        if (mUserLocationReceiver != null) {
```
don't we want to unregister in onPause?

BarAgam replied:
yes we do !resolved

BarAgam replied:
Actually we don't. We spoke about it together and understood its the right place for unregistreation.

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/viewer/cesium/CesiumMapView.java:
```diff
@@ -36,6 +38,14 @@
+     * that occur while JS thread is updating data, that's simultaneously used
+     * by another thread
+     */
+    private SynchronizedDataHolder<Boolean> mIsGGMapReadySynchronized;
```
why do you use this mIsGGMapReadySynchronized?
why dont use just callback. it will be much simpler.

BarAgam replied:
Both calback and a simple get method are available.
I thought I would need the get method performance-wise but it turned out I didn't.
This code can be deleted since It's not used, but it's quite common to have both an "event"/"listener"/callback to state changes AND a direct get method for that state.

barmako committed:
CR fixes:
    Using configurable values to set GGMapView initial extent.
    Resolved race-condition between GGMapView onReady event and it's listener attachment on onCreateView.
    Removed redundant documentation.
